// testcafe/lib/client-functions/selectors/add-api.js
'use strict';

exports.__esModule = true;

var _defineProperty = require('babel-runtime/core-js/object/define-property');

var _defineProperty2 = _interopRequireDefault(_defineProperty);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

let getSnapshot = (() => {
    var _ref = (0, _asyncToGenerator3.default)(function* (getSelector, callsite) {
        var node = null;
        var selector = getSelector();

        try {
            node = yield selector();
        } catch (err) {
            err.callsite = callsite;
            throw err;
        }

        if (!node) throw new _testRun.CantObtainInfoForElementSpecifiedBySelectorError(callsite);

        return node;
    });

    return function getSnapshot(_x, _x2) {
        return _ref.apply(this, arguments);
    };
})();

exports.addCustomMethods = addCustomMethods;
exports.addAPI = addAPI;

var _lodash = require('lodash');

var _builderSymbol = require('../builder-symbol');

var _builderSymbol2 = _interopRequireDefault(_builderSymbol);

var _snapshotProperties = require('./snapshot-properties');

var _testRun = require('../../errors/test-run');

var _getCallsite = require('../../errors/get-callsite');

var _clientFunctionBuilder = require('../client-function-builder');

var _clientFunctionBuilder2 = _interopRequireDefault(_clientFunctionBuilder);

var _reExecutablePromise = require('../../utils/re-executable-promise');

var _reExecutablePromise2 = _interopRequireDefault(_reExecutablePromise);

var _typeAssertions = require('../../errors/runtime/type-assertions');

var _makeRegExp = require('../../utils/make-reg-exp');

var _makeRegExp2 = _interopRequireDefault(_makeRegExp);

var _selectorTextFilter = require('./selector-text-filter');

var _selectorTextFilter2 = _interopRequireDefault(_selectorTextFilter);

var _selectorAttributeFilter = require('./selector-attribute-filter');

var _selectorAttributeFilter2 = _interopRequireDefault(_selectorAttributeFilter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const SNAPSHOT_PROPERTIES = _snapshotProperties.NODE_SNAPSHOT_PROPERTIES.concat(_snapshotProperties.ELEMENT_SNAPSHOT_PROPERTIES);

var filterNodes = new _clientFunctionBuilder2.default((nodes, filter, querySelectorRoot, originNode, ...filterArgs) => {
    if (typeof filter === 'number') {
        var matchingNode = filter < 0 ? nodes[nodes.length + filter] : nodes[filter];

        return matchingNode ? [matchingNode] : [];
    }

    var result = [];

    if (typeof filter === 'string') {
        // NOTE: we can search for elements only in document or element.
        if (querySelectorRoot.nodeType !== 1 && querySelectorRoot.nodeType !== 9) return null;

        var matching = querySelectorRoot.querySelectorAll(filter);
        var matchingArr = [];

        for (var i = 0; i < matching.length; i++) matchingArr.push(matching[i]);

        filter = node => matchingArr.indexOf(node) > -1;
    }

    if (typeof filter === 'function') {
        for (var j = 0; j < nodes.length; j++) {
            if (filter(nodes[j], j, originNode, ...filterArgs)) result.push(nodes[j]);
        }
    }

    return result;
}).getFunction();

var expandSelectorResults = new _clientFunctionBuilder2.default((selector, populateDerivativeNodes) => {
    var nodes = selector();

    if (!nodes.length) return null;

    var result = [];

    for (var i = 0; i < nodes.length; i++) {
        var derivativeNodes = populateDerivativeNodes(nodes[i]);

        if (derivativeNodes) {
            for (var j = 0; j < derivativeNodes.length; j++) {
                if (result.indexOf(derivativeNodes[j]) < 0) result.push(derivativeNodes[j]);
            }
        }
    }

    return result;
}).getFunction();

function assertAddCustomDOMPropertiesOptions(properties) {
    (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomDOMProperties', '"addCustomDOMProperties" option', properties);

    (0, _keys2.default)(properties).forEach(prop => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'addCustomDOMProperties', `Custom DOM properties method '${prop}'`, properties[prop]);
    });
}

function assertAddCustomMethods(properties, opts) {
    (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomMethods', '"addCustomMethods" option', properties);

    if (opts !== void 0) (0, _typeAssertions.assertType)(_typeAssertions.is.nonNullObject, 'addCustomMethods', '"addCustomMethods" option', opts);

    (0, _keys2.default)(properties).forEach(prop => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.function, 'addCustomMethods', `Custom method '${prop}'`, properties[prop]);
    });
}

function addSnapshotProperties(obj, getSelector, properties) {
    properties.forEach(prop => {
        (0, _defineProperty2.default)(obj, prop, {
            get: () => {
                var callsite = (0, _getCallsite.getCallsiteForMethod)('get');

                return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
                    var snapshot = yield getSnapshot(getSelector, callsite);

                    return snapshot[prop];
                }));
            }
        });
    });
}

function addCustomMethods(obj, getSelector, SelectorBuilder, customMethods) {
    const customMethodProps = customMethods ? (0, _keys2.default)(customMethods) : [];

    customMethodProps.forEach(prop => {
        var _customMethods$prop = customMethods[prop],
            _customMethods$prop$r = _customMethods$prop.returnDOMNodes;
        const returnDOMNodes = _customMethods$prop$r === undefined ? false : _customMethods$prop$r,
              method = _customMethods$prop.method;


        const dependencies = {
            customMethod: method,
            selector: getSelector()
        };

        const callsiteNames = { instantiation: prop };

        if (returnDOMNodes) {
            obj[prop] = (...args) => {
                const selectorFn = () => {
                    /* eslint-disable no-undef */
                    const nodes = selector();

                    return customMethod.apply(customMethod, [nodes].concat(args));
                    /* eslint-enable no-undef */
                };

                return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, () => true, {
                    args,
                    customMethod: method
                });
            };
        } else {
            obj[prop] = new _clientFunctionBuilder2.default((...args) => {
                /* eslint-disable no-undef */
                const node = selector();

                return customMethod.apply(customMethod, [node].concat(args));
                /* eslint-enable no-undef */
            }, { dependencies }, callsiteNames).getFunction();
        }
    });
}

function addSnapshotPropertyShorthands(obj, getSelector, SelectorBuilder, customDOMProperties, customMethods) {
    var properties = SNAPSHOT_PROPERTIES;

    if (customDOMProperties) properties = properties.concat((0, _keys2.default)(customDOMProperties));

    addSnapshotProperties(obj, getSelector, properties);
    addCustomMethods(obj, getSelector, SelectorBuilder, customMethods);

    obj.getStyleProperty = prop => {
        var callsite = (0, _getCallsite.getCallsiteForMethod)('getStyleProperty');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            var snapshot = yield getSnapshot(getSelector, callsite);

            return snapshot.style ? snapshot.style[prop] : void 0;
        }));
    };

    obj.getAttribute = attrName => {
        var callsite = (0, _getCallsite.getCallsiteForMethod)('getAttribute');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            var snapshot = yield getSnapshot(getSelector, callsite);

            return snapshot.attributes ? snapshot.attributes[attrName] : void 0;
        }));
    };

    obj.hasAttribute = attrName => {
        var callsite = (0, _getCallsite.getCallsiteForMethod)('hasAttribute');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            var snapshot = yield getSnapshot(getSelector, callsite);

            return snapshot.attributes ? snapshot.attributes.hasOwnProperty(attrName) : false;
        }));
    };

    obj.getBoundingClientRectProperty = prop => {
        var callsite = (0, _getCallsite.getCallsiteForMethod)('getBoundingClientRectProperty');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            var snapshot = yield getSnapshot(getSelector, callsite);

            return snapshot.boundingClientRect ? snapshot.boundingClientRect[prop] : void 0;
        }));
    };

    obj.hasClass = name => {
        var callsite = (0, _getCallsite.getCallsiteForMethod)('hasClass');

        return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
            var snapshot = yield getSnapshot(getSelector, callsite);

            return snapshot.classNames ? snapshot.classNames.indexOf(name) > -1 : false;
        }));
    };
}

function createCounter(getSelector, SelectorBuilder) {
    var builder = new SelectorBuilder(getSelector(), { counterMode: true }, { instantiation: 'Selector' });
    var counter = builder.getFunction();
    var callsite = (0, _getCallsite.getCallsiteForMethod)('get');

    return (0, _asyncToGenerator3.default)(function* () {
        try {
            return yield counter();
        } catch (err) {
            err.callsite = callsite;
            throw err;
        }
    });
}

function addCounterProperties(obj, getSelector, SelectorBuilder) {
    Object.defineProperty(obj, 'count', {
        get: () => {
            var counter = createCounter(getSelector, SelectorBuilder);

            return _reExecutablePromise2.default.fromFn(() => counter());
        }
    });

    Object.defineProperty(obj, 'exists', {
        get: () => {
            var counter = createCounter(getSelector, SelectorBuilder);

            return _reExecutablePromise2.default.fromFn((0, _asyncToGenerator3.default)(function* () {
                return (yield counter()) > 0;
            }));
        }
    });
}

function convertFilterToClientFunctionIfNecessary(callsiteName, filter, dependencies) {
    if (typeof filter === 'function') {
        var builder = filter[_builderSymbol2.default];
        var fn = builder ? builder.fn : filter;
        var options = builder ? (0, _lodash.assign)({}, builder.options, { dependencies }) : { dependencies };

        return new _clientFunctionBuilder2.default(fn, options, { instantiation: callsiteName }).getFunction();
    }

    return filter;
}

function createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, additionalDependencies) {
    var collectionModeSelectorBuilder = new SelectorBuilder(getSelector(), { collectionMode: true });
    var customDOMProperties = collectionModeSelectorBuilder.options.customDOMProperties;
    var customMethods = collectionModeSelectorBuilder.options.customMethods;

    var dependencies = {
        selector: collectionModeSelectorBuilder.getFunction(),
        filter: filter,
        filterNodes: filterNodes
    };

    var _collectionModeSelect = collectionModeSelectorBuilder.options,
        boundTestRun = _collectionModeSelect.boundTestRun,
        timeout = _collectionModeSelect.timeout,
        visibilityCheck = _collectionModeSelect.visibilityCheck;


    dependencies = (0, _lodash.assign)(dependencies, additionalDependencies);

    var builder = new SelectorBuilder(selectorFn, {
        dependencies,
        customDOMProperties,
        customMethods,
        boundTestRun,
        timeout,
        visibilityCheck
    }, { instantiation: 'Selector' });

    return builder.getFunction();
}

var filterByText = convertFilterToClientFunctionIfNecessary('filter', _selectorTextFilter2.default);
var filterByAttr = convertFilterToClientFunctionIfNecessary('filter', _selectorAttributeFilter2.default);

function ensureRegExpContext(str) {
    // NOTE: if a regexp is created in a separate context (via the 'vm' module) we
    // should wrap it with new RegExp() to make the `instanceof RegExp` check successful.
    if (typeof str !== 'string' && !(str instanceof RegExp)) return new RegExp(str);

    return str;
}

function addFilterMethods(obj, getSelector, SelectorBuilder) {
    obj.nth = index => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.number, 'nth', '"index" argument', index);

        var builder = new SelectorBuilder(getSelector(), { index: index }, { instantiation: 'Selector' });

        return builder.getFunction();
    };

    obj.withText = text => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withText', '"text" argument', text);

        text = ensureRegExpContext(text);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            var nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, textRe);
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filterByText, {
            textRe: (0, _makeRegExp2.default)(text)
        });
    };

    obj.withExactText = text => {
        (0, _typeAssertions.assertType)(_typeAssertions.is.string, 'withExactText', '"text" argument', text);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            var nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, exactText);
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filterByText, {
            exactText: text
        });
    };

    obj.withAttribute = (attrName, attrValue) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withAttribute', '"attrName" argument', attrName);

        attrName = ensureRegExpContext(attrName);

        if (attrValue !== void 0) {
            (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.regExp], 'withAttribute', '"attrValue" argument', attrValue);
            attrValue = ensureRegExpContext(attrValue);
        }

        var selectorFn = () => {
            /* eslint-disable no-undef */
            var nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0, attrName, attrValue);
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filterByAttr, {
            attrName,
            attrValue
        });
    };

    obj.filter = (filter, dependencies) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function], 'filter', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('filter', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            var nodes = selector();

            if (!nodes.length) return null;

            return filterNodes(nodes, filter, document, void 0);
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter);
    };

    obj.filterVisible = () => {
        const builder = new SelectorBuilder(getSelector(), { filterVisible: true }, { instantiation: 'Selector' });

        return builder.getFunction();
    };

    obj.filterHidden = () => {
        const builder = new SelectorBuilder(getSelector(), { filterHidden: true }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addCustomDOMPropertiesMethod(obj, getSelector, SelectorBuilder) {
    obj.addCustomDOMProperties = customDOMProperties => {
        assertAddCustomDOMPropertiesOptions(customDOMProperties);

        var builder = new SelectorBuilder(getSelector(), { customDOMProperties }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addCustomMethodsMethod(obj, getSelector, SelectorBuilder) {
    obj.addCustomMethods = function (methods, opts) {
        assertAddCustomMethods(methods, opts);

        const customMethods = {};

        (0, _keys2.default)(methods).forEach(methodName => {
            customMethods[methodName] = {
                method: methods[methodName],
                returnDOMNodes: opts && !!opts.returnDOMNodes
            };
        });

        const builder = new SelectorBuilder(getSelector(), { customMethods }, { instantiation: 'Selector' });

        return builder.getFunction();
    };
}

function addHierarchicalSelectors(obj, getSelector, SelectorBuilder) {
    // Find
    obj.find = (filter, dependencies) => {
        (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function], 'find', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                if (typeof filter === 'string') {
                    return typeof node.querySelectorAll === 'function' ? node.querySelectorAll(filter) : null;
                }

                var results = [];

                var visitNode = currentNode => {
                    var cnLength = currentNode.childNodes.length;

                    for (var i = 0; i < cnLength; i++) {
                        var child = currentNode.childNodes[i];

                        results.push(child);

                        visitNode(child);
                    }
                };

                visitNode(node);

                return filterNodes(results, filter, null, node);
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };

    // Parent
    obj.parent = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'parent', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                var parents = [];

                for (var parent = node.parentNode; parent; parent = parent.parentNode) parents.push(parent);

                return filter !== void 0 ? filterNodes(parents, filter, document, node) : parents;
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };

    // Child
    obj.child = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'child', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                var childElements = [];
                var cnLength = node.childNodes.length;

                for (var i = 0; i < cnLength; i++) {
                    var child = node.childNodes[i];

                    if (child.nodeType === 1) childElements.push(child);
                }

                return filter !== void 0 ? filterNodes(childElements, filter, node, node) : childElements;
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };

    // Sibling
    obj.sibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'sibling', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                var parent = node.parentNode;

                if (!parent) return null;

                var siblings = [];
                var cnLength = parent.childNodes.length;

                for (var i = 0; i < cnLength; i++) {
                    var child = parent.childNodes[i];

                    if (child.nodeType === 1 && child !== node) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };

    // Next sibling
    obj.nextSibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'nextSibling', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                var parent = node.parentNode;

                if (!parent) return null;

                var siblings = [];
                var cnLength = parent.childNodes.length;
                var afterNode = false;

                for (var i = 0; i < cnLength; i++) {
                    var child = parent.childNodes[i];

                    if (child === node) afterNode = true;else if (afterNode && child.nodeType === 1) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };

    // Prev sibling
    obj.prevSibling = (filter, dependencies) => {
        if (filter !== void 0) (0, _typeAssertions.assertType)([_typeAssertions.is.string, _typeAssertions.is.function, _typeAssertions.is.number], 'prevSibling', '"filter" argument', filter);

        filter = convertFilterToClientFunctionIfNecessary('find', filter, dependencies);

        var selectorFn = () => {
            /* eslint-disable no-undef */
            return expandSelectorResults(selector, node => {
                var parent = node.parentNode;

                if (!parent) return null;

                var siblings = [];
                var cnLength = parent.childNodes.length;

                for (var i = 0; i < cnLength; i++) {
                    var child = parent.childNodes[i];

                    if (child === node) break;

                    if (child.nodeType === 1) siblings.push(child);
                }

                return filter !== void 0 ? filterNodes(siblings, filter, parent, node) : siblings;
            });
            /* eslint-enable no-undef */
        };

        return createDerivativeSelectorWithFilter(getSelector, SelectorBuilder, selectorFn, filter, { expandSelectorResults });
    };
}

function addAPI(obj, getSelector, SelectorBuilder, customDOMProperties, customMethods) {
    addSnapshotPropertyShorthands(obj, getSelector, SelectorBuilder, customDOMProperties, customMethods);
    addCustomDOMPropertiesMethod(obj, getSelector, SelectorBuilder);
    addCustomMethodsMethod(obj, getSelector, SelectorBuilder);
    addFilterMethods(obj, getSelector, SelectorBuilder);
    addHierarchicalSelectors(obj, getSelector, SelectorBuilder);
    addCounterProperties(obj, getSelector, SelectorBuilder);
}
